File: packages/execution-go/auth/servicejwt.go
package auth

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"strings"
	"time"
)

type Claims struct {
	Issuer    string `json:"iss"`
	Audience  string `json:"aud"`
	Service   string `json:"service"`
	Env       string `json:"env"`
	RequestID string `json:"requestId"`
	IssuedAt  int64  `json:"iat"`
	ExpiresAt int64  `json:"exp"`
}

func Sign(claims Claims, secret string) (string, error) {
	if secret == "" {
		return "", errors.New("service token secret is required")
	}
	header, err := json.Marshal(map[string]string{"alg": "HS256", "typ": "JWT"})
	if err != nil {
		return "", err
	}
	payload, err := json.Marshal(claims)
	if err != nil {
		return "", err
	}
	unsigned := base64.RawURLEncoding.EncodeToString(header) + "." +
		base64.RawURLEncoding.EncodeToString(payload)
	mac := hmac.New(sha256.New, []byte(secret))
	_, _ = mac.Write([]byte(unsigned))
	return unsigned + "." + base64.RawURLEncoding.EncodeToString(mac.Sum(nil)), nil
}

func Verify(token, secret, issuer, audience, environment string, now time.Time) (Claims, error) {
	var claims Claims
	parts := strings.Split(token, ".")
	if len(parts) != 3 || secret == "" {
		return claims, errors.New("invalid service token")
	}
	mac := hmac.New(sha256.New, []byte(secret))
	_, _ = mac.Write([]byte(parts[0] + "." + parts[1]))
	expected := mac.Sum(nil)
	signature, err := base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil || !hmac.Equal(signature, expected) {
		return claims, errors.New("invalid service token signature")
	}
	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil || json.Unmarshal(payload, &claims) != nil {
		return claims, errors.New("invalid service token claims")
	}
	if claims.Issuer != issuer || claims.Audience != audience || claims.Env != environment {
		return claims, errors.New("service token scope mismatch")
	}
	unix := now.Unix()
	if claims.ExpiresAt <= unix || claims.IssuedAt > unix+30 || claims.ExpiresAt-claims.IssuedAt > 300 {
		return claims, errors.New("service token expired or lifetime invalid")
	}
	return claims, nil
}

File: lib/command-center/execution/service-jwt.ts
import { createHmac, timingSafeEqual } from "node:crypto";

export interface ServiceJWTClaims {
  iss: string;
  aud: string;
  service: string;
  env: string;
  requestId: string;
  iat: number;
  exp: number;
}

function encode(value: object): string {
  return Buffer.from(JSON.stringify(value)).toString("base64url");
}

export function signServiceJWT(
  claims: ServiceJWTClaims,
  secret: string,
): string {
  if (!secret) throw new Error("Execution service JWT secret is required.");
  const unsigned = `${encode({ alg: "HS256", typ: "JWT" })}.${encode(claims)}`;
  const signature = createHmac("sha256", secret)
    .update(unsigned)
    .digest("base64url");
  return `${unsigned}.${signature}`;
}

export function verifyServiceJWT(
  token: string,
  expected: {
    issuer: string;
    audience: string;
    environment: string;
    secret: string;
    now?: number;
  },
): ServiceJWTClaims {
  const parts = token.split(".");
  if (parts.length !== 3) throw new Error("Invalid service token.");
  const unsigned = `${parts[0]}.${parts[1]}`;
  const actual = Buffer.from(parts[2], "base64url");
  const wanted = createHmac("sha256", expected.secret)
    .update(unsigned)
    .digest();
  if (actual.length !== wanted.length || !timingSafeEqual(actual, wanted)) {
    throw new Error("Invalid service token signature.");
  }
  const parsed: unknown = JSON.parse(
    Buffer.from(parts[1], "base64url").toString("utf8"),
  );
  if (!isServiceClaims(parsed)) throw new Error("Invalid service token claims.");
  const now = expected.now ?? Math.floor(Date.now() / 1000);
  if (
    parsed.iss !== expected.issuer ||
    parsed.aud !== expected.audience ||
    parsed.env !== expected.environment ||
    parsed.exp <= now ||
    parsed.iat > now + 30 ||
    parsed.exp - parsed.iat > 300
  ) {
    throw new Error("Service token scope or lifetime is invalid.");
  }
  return parsed;
}

function isServiceClaims(value: unknown): value is ServiceJWTClaims {
  if (!value || typeof value !== "object") return false;
  const claims = value as Record<string, unknown>;
  return (
    typeof claims.iss === "string" &&
    typeof claims.aud === "string" &&
    typeof claims.service === "string" &&
    typeof claims.env === "string" &&
    typeof claims.requestId === "string" &&
    typeof claims.iat === "number" &&
    typeof claims.exp === "number"
  );
}
